# Diamond Prices
## Give an overview of the dataset structure by answering those questions:

In [125]:
import numpy as np
import pandas as pd
from pandas import DataFrame

df = pd.read_csv('diamonds.csv')
df.describe() # list some statistics for the number features in the dataset


carat         depth         table         price             x  \
count  53940.000000  53940.000000  53940.000000  53940.000000  53940.000000   
mean       0.797940     61.749405     57.457184   3932.799722      5.731157   
std        0.474011      1.432621      2.234491   3989.439738      1.121761   
min        0.200000     43.000000     43.000000    326.000000      0.000000   
25%        0.400000     61.000000     56.000000    950.000000      4.710000   
50%        0.700000     61.800000     57.000000   2401.000000      5.700000   
75%        1.040000     62.500000     59.000000   5324.250000      6.540000   
max        5.010000     79.000000     95.000000  18823.000000     10.740000   

                  y             z  
count  53940.000000  53940.000000  
mean       5.734526      3.538734  
std        1.142135      0.705699  
min        0.000000      0.000000  
25%        4.720000      2.910000  
50%        5.710000      3.530000  
75%        6.540000      4.040000  
max       58.900000     31.800000

### How many samples and features are in the dataset?

In [126]:
print('Number of samples: ')
len(df)

Number of samples: 


53940

In [127]:
print('Number of features: ')
len(df.columns)

Number of features: 


10

### What are the feature data types?

In [128]:
df.dtypes

carat      float64
cut         object
color       object
clarity     object
depth      float64
table      float64
price        int64
x          float64
y          float64
z          float64
dtype: object

### Are diamonds balanced across color, cut and clarity? (Hint: roughly 1:1 means balanced, e.g. 1:2 is a “1:2 imbalance”)

In [129]:
""""
Calculate the class distribution ratio of a feature by using the feature with the least samples as the basis for the calculation
"""
def ratio(df: DataFrame, feature: str):
    g = df.groupby([feature]).size().to_frame(name = "samples")
    min_samples = g.iloc[:, 0].min()
    g['ratio'] = round(g.iloc[:, 0] / min_samples, 1)
    return g

In [130]:
ratio(df, "color")

samples  ratio
color                
D         6775    2.4
E         9797    3.5
F         9542    3.4
G        11292    4.0
H         8304    3.0
I         5422    1.9
J         2808    1.0

Color has an imbalance of 1:4 from class J to F and over half the classes have at least an imbalance of 1:3 when compared to J

In [131]:
ratio(df, "cut")

samples  ratio
cut                      
Fair          1610    1.0
Good          4906    3.0
Ideal        21551   13.4
Premium      13791    8.6
Very Good    12082    7.5

Cut has an even higher imbalance of 1:13.4 between Fair and Ideal. in general cut is very unbalanced

In [132]:
ratio(df, "clarity")

samples  ratio
clarity                
I1           741    1.0
IF          1790    2.4
SI1        13065   17.6
SI2         9194   12.4
VS1         8171   11.0
VS2        12258   16.5
VVS1        3655    4.9
VVS2        5066    6.8

Clarity has the worst imbalance of 1:17.6 between I1 and SI1 and apart from that is the most imbalanced feature of the three